design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+Environment bootstrapping doesn't have to be performed by the development team, and does not require deep knowledge of the application structure, besides the set of accounts and regions into which the app needs to be deployed.
+
+The current implementation only provisions an S3 bucket, but in order to be able to continuously deploy CDK stacks that use asses, we will need the following resources:
+
+For publishing:
+
+* **S3 Bucket**: for file asset and CloudFormation templates
+* **ECR Repository**: for docker image assets
+* **Publishing Role**: IAM role trusted by the deployment account, and allows publishing to the S3 bucket and the ECR repository.
+
+For deployment:
+
+* **CloudFormation Role**: IAM role which allows the CloudFormation service principal to deploy stacks into the environment (this role usually has administrative privileges).
+* **Deployment Role**: IAM role which allows anyone from the deployment account to create, update and describe CloudFormation change sets and s"pass" the CloudFormation role (`iam:PassRole`).
+
+To accommodate these requirements we will make the following changes to how `cdk bootstrap` works:
+
+1. Extend the bootstrap stack to include these resources.
+2. Use explicit convention-based physical names for all resources.
+3. Allow specifying a list of *trusted accounts* from which the CloudFormation service principal can deploy to this account.
+4. Allow specifying a list of accounts from which principals can deploy to this account from the CLI.
+5. Allow specifying the managed policy to use for the deployment role (mostly it will be the administrator managed policy).
+6. Allow specifying an optional qualifier for the physical names of all resources to address bucket hijacking concerns and allow multiple bootstraps to the same environment for whatever reason.
+
+**Resource Names**
+
+We need to be able to synthesize asset locations into the templates for consumption and publishing. We also need to be able to assume a role in order to be able to publish to the environment.
+
+This means that we cannot rely on CloudFormation physical name generation since it requires accessing the account in order to resolve these names.
+
+We will employ a naming convention which encodes `account`, `region` and an optional `qualifier` (such as `cdk-account-region[-qualifier]-xxxx`). This is because not all AWS resource names are environment-local: IAM roles are account-wide and S3 buckets are global.
+
+It is important that we do not rely on hashing or parsing account and region in order to be able to support environment-agnostic stacks (in which case "account" resolves to `{ "Ref": "AWS::AccountId" }`, etc.
+
+In order to address the risk of S3 bucket hijacking, we need to be ale to support an optional `qualifier` postfix. This means that we need to allow users to specify this qualifier when they define the Stack's `env`. Perhaps we need to encode this into `aws://account/region[/qualifier]`
```

Comment:
Just a typo... assume it's meant to be 'able'.

Reply:
Thanks